From 6bc1cf09a90a92992e00faa083f8741c7dfa060d Mon Sep 17 00:00:00 2001
From: Harm Hanemaaijer <fgenfb@yahoo.com>
Date: Sat, 21 Feb 2015 08:51:57 +0100
Subject: [PATCH 1/2] Disable use of DMA fb_copyarea for console scrolling on
 Pi 2

By default, fb_copyarea is accelerated using DMA on the Raspberry Pi
and this brings improved performance for console text scrolling on
first generation Raspberry Pi models.

However, on the Raspberry Pi 2, the situation is less clear-cut;
disabling the use of accelerated fb_copyarea, which triggers the use
of optimized redraw when scrolling, results in clearly better
performance. The gains are greatest at 16bpp and high display
resolutions like 1920x1080. At 1920x1080, using a console text
throughput test, the performance gain ranges from 11% at 32bpp to 26%
at 16bpp. When a concurrent CPU-intensive task is introduced, the gain
 increases from a range of 16% at 32bpp to 32% at 16bpp.

When using Raspberry Pi 2 with aggressive overclocking settings (1100
MHz CPU, 550 MHz L2 cache and 483 MHz SDRAM), the advantage dimishes,
with the gain ranging from -5% at 32bpp to 11% at 16bpp. When a
concurrent CPU-intensive task is introduced, the gain increases from a
range of 3% at 32bpp to 19% at 16bpp.

At the lower resolution of 1280x1024, the performance gain is less. On
an overclocked Raspberry Pi 2, the performance gain ranges from -6% at
32bpp to 7% at 16bpp. When a concurrent CPU-intensive task is
introduced, the gain increases from a range of -1% at 32bpp to 18% at
16bpp.

To explain why redraw can be preferable to DMA for scrolling, consider
that copyarea uses approximately twice the amount of memory bandwidth
because of the need to read the source region. Redraw only writes to
the destination region, saving bandwidth. On the Raspberry Pi 2, with
its increased CPU speed and parallellism, memory bandwidth is more of
a bottleneck and optimized software redraw performs better.

An analogy between the X server and the kernel can also be put
forward. The Raspberry Pi 2 benefits from X server graphics operations
being executed on one CPU, leaving other CPUs free for applications
use. In a similar way using software redraw in the kernel works better
on the Raspberry Pi 2.

Care should be taken that accelerated fb_copyarea remains in use for
real screen area copy requests in the X server as provided by the
fbturbo driver.

This commit adds the FBINFO_HWACCEL_DISABLED flag to the fb flags used
internally in the kernel, which has the effect that accelerated
fb_copyarea is no longer used for scrolling. Accelerated fb_copyarea
can still be taken advantage of by other applications, such as the
Xorg fbturbo driver.

Signed-off-by: Harm Hanemaaijer <fgenfb@yahoo.com>
---
 drivers/video/fbdev/bcm2708_fb.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/video/fbdev/bcm2708_fb.c b/drivers/video/fbdev/bcm2708_fb.c
index a443666..81dfb32 100644
--- a/drivers/video/fbdev/bcm2708_fb.c
+++ b/drivers/video/fbdev/bcm2708_fb.c
@@ -639,7 +639,14 @@ static int bcm2708_fb_register(struct bcm2708_fb *fb)
 		fb->dma = dma;
 	}
 	fb->fb.fbops = &bcm2708_fb_ops;
+#if __LINUX_ARM_ARCH__ >= 7
+        /* Disable used of accelerated fb_copyarea for scrolling on
+         * on Raspberry Pi 2. */
+	fb->fb.flags = FBINFO_FLAG_DEFAULT | FBINFO_HWACCEL_COPYAREA |
+		FBINFO_HWACCEL_DISABLED;
+#else
 	fb->fb.flags = FBINFO_FLAG_DEFAULT | FBINFO_HWACCEL_COPYAREA;
+#endif
 	fb->fb.pseudo_palette = fb->cmap;
 
 	strncpy(fb->fb.fix.id, bcm2708_name, sizeof(fb->fb.fix.id));
-- 
1.9.1

